internal/component/prometheus/remote/queue/serialization/serializer.go

```diff
@@ -0,0 +1,223 @@
+			return actor.WorkerEnd
+		}
+		s.maxItemsBeforeFlush = int(cfg.MaxSignalsInBatch)
+		s.flushFrequency = cfg.FlushFrequency
+		return actor.WorkerContinue
+	default:
+	}
+
+	select {
+	case <-ctx.Done():
+		return actor.WorkerEnd
+	case item, ok := <-s.inbox.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		err := s.Append(ctx, item)
+		if err != nil {
+			level.Error(s.logger).Log("msg", "unable to append to serializer", "err", err)
+		}
+		return actor.WorkerContinue
+	case item, ok := <-s.metaInbox.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		err := s.AppendMetadata(ctx, item)
+		if err != nil {
+			level.Error(s.logger).Log("msg", "unable to append metadata to serializer", "err", err)
+		}
+		return actor.WorkerContinue
+	case <-s.flushTestTimer.C:
+		if time.Since(s.lastFlush) > s.flushFrequency {
+			err := s.store(ctx)
+			if err != nil {
+				level.Error(s.logger).Log("msg", "unable to store data", "err", err)
+			}
+		}
+		return actor.WorkerContinue
+	}
+}
+
+func (s *serializer) AppendMetadata(ctx actor.Context, data *types.TimeSeriesBinary) error {
+	s.meta = append(s.meta, data)
+	// If we would go over the max size then send, or if we have hit the flush duration then send.
+	if len(s.meta)+len(s.series) >= s.maxItemsBeforeFlush {
+		return s.store(ctx)
+	} else if time.Since(s.lastFlush) > s.flushFrequency {
+		return s.store(ctx)
+	}
+	return nil
+}
+
+func (s *serializer) Append(ctx actor.Context, data *types.TimeSeriesBinary) error {
+	s.series = append(s.series, data)
+	// If we would go over the max size then send, or if we have hit the flush duration then send.
+	if len(s.meta)+len(s.series) >= s.maxItemsBeforeFlush {
+		return s.store(ctx)
+	} else if time.Since(s.lastFlush) > s.flushFrequency {
+		return s.store(ctx)
+	}
+	return nil
+}
+
+func (s *serializer) store(ctx actor.Context) error {
+	var err error
+	defer func() {
+		s.lastFlush = time.Now()
+	}()
+	// Do nothing if there is nothing.
+	if len(s.series) == 0 && len(s.meta) == 0 {
+		return nil
+	}
+	group := &types.SeriesGroup{
+		Series:   make([]*types.TimeSeriesBinary, len(s.series)),
+		Metadata: make([]*types.TimeSeriesBinary, len(s.meta)),
+	}
+	defer func() {
+		s.storeStats(err)
+		// Return series to the pool, this is key to reducing allocs.
+		types.PutTimeSeriesBinarySlice(s.series)
+		types.PutTimeSeriesBinarySlice(s.meta)
+		s.series = s.series[:0]
+		s.meta = s.series[:0]
+	}()
+
+	// This maps strings to index position in a slice. This is doing to reduce the file size of the data.
+	strMapToInt := make(map[string]uint32)
+	for i, ts := range s.series {
+		ts.FillLabelMapping(strMapToInt)
+		group.Series[i] = ts
+	}
+	for i, ts := range s.meta {
+		ts.FillLabelMapping(strMapToInt)
+		group.Metadata[i] = ts
+	}
+
+	stringsSlice := make([]string, len(strMapToInt))
+	for stringValue, index := range strMapToInt {
+		stringsSlice[index] = stringValue
+	}
+	group.Strings = stringsSlice
+
+	buf, err := group.MarshalMsg(s.msgpBuffer)
```

Comment:
Sooo... is it worth it to do the dictionary stuff? I guess yes, but on the other hand I know that compression algos would do something similar automatically, snappy [can refer to previous part of the data to reduce repetition](https://en.wikipedia.org/wiki/Snappy_(compression)#Example_of_a_compressed_stream).

Reply:
I hadnt tested since I used msgp so recreated a test, 100k metrics with 10 labels apiece. The name being relatively constant the the value being random. The mapping approach saves roughly 33%, the less random the data is the better savings. If we reduce the random string generation to only 5 characters its half the size of the raw dictionary.

```
//go:generate msgp
package main

import (
	"fmt"
	"math/rand"

	"github.com/golang/snappy"
)

func main() {
	metrics := make([]map[string]string, 0)
	// 100k metrics with 10 labels each
	for i := 0; i < 100_000; i++ {
		metrics = append(metrics, getLabels())

	}

	ss := &StringString{Labels: metrics}
	bb, err := ss.MarshalMsg(nil)
	if err != nil {
		panic(err)
	}
	out := snappy.Encode(nil, bb)
	println(fmt.Printf("dictionary based is %d bytes", len(out)))

	ib := &IndexBased{
		String: make([]string, 0),
		Labels: make([]map[uint32]uint32, 0),
	}
	alignIndexBased(ib, metrics)
	bb, err = ib.MarshalMsg(nil)
	if err != nil {
		panic(err)
	}
	out = snappy.Encode(nil, bb)
	println(fmt.Printf("index based is %d bytes", len(out)))

}

type StringString struct {
	Labels []map[string]string
}

type IndexBased struct {
	String []string
	Labels []map[uint32]uint32
}

func alignIndexBased(ib *IndexBased, strings []map[string]string) {
	index := 0
	stringsList := make(map[string]int)
	for _, metric := range strings {
		mapping := make(map[uint32]uint32)
		for k, v := range metric {
			keyIndex, ok := stringsList[k]
			if !ok {
				stringsList[k] = index
				ib.String = append(ib.String, k)
				keyIndex = index
				index++
			}
			valIndex, ok := stringsList[v]
			if !ok {
				stringsList[k] = index
				ib.String = append(ib.String, v)
				valIndex = index
				index++
			}
			mapping[uint32(keyIndex)] = uint32(valIndex)
		}
		ib.Labels = append(ib.Labels, mapping)

	}

}

func getLabels() map[string]string {
	retLbls := make(map[string]string, 0)
	for i := 0; i < 10; i++ {
		retLbls[fmt.Sprintf("label_%d", i)] = randString()
	}
	return retLbls
}

var letterRunes = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")

func randString() string {
	b := make([]rune, rand.Intn(20))
	for i := range b {
		b[i] = letterRunes[rand.Intn(len(letterRunes))]
	}
	return string(b)
}
 
```

Reply:
In general the index based is never worse, and IMO in many cases is 60% of the size of the pure string based.
Results from the above test, changing out the letterRunes to small sets.

// 5 char long really random
// 5371616
// 4732108

// 5 long half random
// 5050060
// 3929185

// 5 long quarter random
// 4455979
// 2918973

Reply:
The lower the cardinality the better but even in worse case its not terrible.